### PR TITLE
rpc: implement internal `debug_` API namespace functions

### PR DESCRIPTION
**Pull request by hanchon**
Ref: #71 (This PR is 1/3 of the issue)

## Description

This PR is the first one to start implementing the `_debug` namespace.
The `_debug` namespace has 3 parts:
* Internal debug: Profiling functions. This PR.
* Logger configuration: Allows the user the change the logger settings, for example the log level. This needs to be implemented in another PR.
* Trace calls: This calls return blockchain information (Blocks and Transactions). This needs to be implemented to support etherscan. 

______

For contributor use:

- [X] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work. (This was internally discussed with @fedekunze)
- [X] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
(Note: we based the code on the go-etherum codebase, it test some functionality from the runtime lib, but not the API endpoint. For example: https://github.com/ethereum/go-ethereum/blob/master/metrics/runtime_test.go, we can add these tests if they are needed)
- [X] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [X] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [X] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [X] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: ethereum/rpc/namespaces/debug/internal.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+	runtime.SetBlockProfileRate(1)
+	defer runtime.SetBlockProfileRate(0)
+
+	time.Sleep(time.Duration(nsec) * time.Second)
+	return writeProfile("block", file, a.logger)
+}
+
+// CpuProfile turns on CPU profiling for nsec seconds and writes
+// profile data to file.
+func (a *InternalAPI) CpuProfile(file string, nsec uint) error {
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/debug/internal.go:59]:
```suggestion
func (a *InternalAPI) CpuProfile(file string, nsec uint) error { // nolint: golint
```

File: ethereum/rpc/namespaces/debug/internal.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	a.logger.Debug("debug_cpuProfile", "file", file, "nsec", nsec)
+	if err := a.StartCPUProfile(file); err != nil {
+		return err
+	}
+	time.Sleep(time.Duration(nsec) * time.Second)
+	a.StopCPUProfile()
+	return nil
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/debug/internal.go:66]:
```suggestion
	return a.StopCPUProfile()
```

File: ethereum/rpc/namespaces/debug/internal.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+// GoTrace turns on tracing for nsec seconds and writes
+// trace data to file.
+func (a *InternalAPI) GoTrace(file string, nsec uint) error {
+	a.logger.Debug("debug_goTrace", "file", file, "nsec", nsec)
+	if err := a.StartGoTrace(file); err != nil {
+		return err
+	}
+	time.Sleep(time.Duration(nsec) * time.Second)
+	a.StopGoTrace()
+	return nil
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/debug/internal.go:86]:
```suggestion
	return a.StopGoTrace()
```

File: ethereum/rpc/namespaces/debug/internal.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+func (a *InternalAPI) SetBlockProfileRate(rate int) {
+	a.logger.Debug("debug_setBlockProfileRate", "rate", rate)
+	runtime.SetBlockProfileRate(rate)
+}
+
+// Stacks returns a printed representation of the stacks of all goroutines.
+func (a *InternalAPI) Stacks() string {
+	a.logger.Debug("debug_stacks")
+	buf := new(bytes.Buffer)
+	pprof.Lookup("goroutine").WriteTo(buf, 2)
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/debug/internal.go:108]: log error

File: ethereum/rpc/namespaces/debug/internal.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	return buf.String()
+}
+
+// StartCPUProfile turns on CPU profiling, writing to the given file.
+func (a *InternalAPI) StartCPUProfile(file string) error {
+	a.logger.Debug("debug_startCPUProfile", "file", file)
+	a.handler.mu.Lock()
+	defer a.handler.mu.Unlock()
+
+	if isCPUProfileConfigurationActivated(a.ctx) {
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/debug/internal.go:118]: use `switch` instead

File: ethereum/rpc/namespaces/debug/internal.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+	}
+}
+
+// StopCPUProfile stops an ongoing CPU profile.
+func (a *InternalAPI) StopCPUProfile() error {
+	a.logger.Debug("debug_stopCPUProfile")
+	a.handler.mu.Lock()
+	defer a.handler.mu.Unlock()
+
+	if isCPUProfileConfigurationActivated(a.ctx) {
```
Comment (fedekunze) [inline ethereum/rpc/namespaces/debug/internal.go:150]: ditto

**Review by fedekunze — COMMENTED** [review]
Looks great! Minor stylistic comments only

File: ethereum/rpc/namespaces/debug/internal.go (hunk 6 above)
Comment (ramacarlucho) [reply ethereum/rpc/namespaces/debug/internal.go]: Implemented all suggested changes

**Review by fedekunze — APPROVED** [review]
ACK

**Pull request merged**